packages/storage-plus/src/indexed_map.rs

```diff
@@ -258,15 +294,123 @@ mod test {
-        assert_eq!(pk.to_vec(), aged.0);
-        assert_eq!(data, aged.1);
+        assert_eq!(pk1.to_vec(), aged.0);
+        assert_eq!(data1, aged.1);
 
         // no match on wrong age
         let too_old = index_int(43);
         let aged = map.idx.age.item(&store, &too_old).unwrap();
         assert_eq!(None, aged);
     }
 
+    #[test]
+    fn range_simple_key_by_multi_index() {
+        let mut store = MockStorage::new();
+        let mut map = build_map();
+
+        // save data
+        let data1 = Data {
+            name: "Maria".to_string(),
+            age: 42,
+        };
+        let pk: &[u8] = b"5627";
+        map.save(&mut store, pk, &data1).unwrap();
+
+        let data2 = Data {
+            name: "Juan".to_string(),
+            age: 13,
+        };
+        let pk: &[u8] = b"5628";
+        map.save(&mut store, pk, &data2).unwrap();
+
+        let data3 = Data {
+            name: "Maria".to_string(),
+            age: 24,
+        };
+        let pk: &[u8] = b"5629";
+        map.save(&mut store, pk, &data3).unwrap();
+
+        let data4 = Data {
+            name: "Maria Luisa".to_string(),
+            age: 12,
+        };
+        let pk: &[u8] = b"5630";
+        map.save(&mut store, pk, &data4).unwrap();
+
+        let marias: StdResult<Vec<_>> = map
+            .idx
+            .name
+            .range(
+                &store,
+                Some(Bound::Inclusive(to_length_prefixed(b"Maria"))),
+                None,
+                Order::Ascending,
+            )
+            .collect();
+        let count = marias.unwrap().len();
+        assert_eq!(3, count);
+    }
+
+    #[test]
+    fn range_composite_key_by_multi_index() {
+        let mut store = MockStorage::new();
+
+        let indexes = DataCompositeMultiIndex {
+            name_age: MultiIndex::new(|d| index_tuple(&d.name, d.age), "data", "data__name_age"),
+        };
+        let mut map = IndexedMap::new("data", indexes);
+
+        // save data
+        let data1 = Data {
+            name: "Maria".to_string(),
+            age: 42,
+        };
+        let pk: &[u8] = b"5627";
+        map.save(&mut store, pk, &data1).unwrap();
+
+        let data2 = Data {
+            name: "Juan".to_string(),
+            age: 13,
+        };
+        let pk: &[u8] = b"5628";
+        map.save(&mut store, pk, &data2).unwrap();
+
+        let data3 = Data {
+            name: "Maria".to_string(),
+            age: 24,
+        };
+        let pk: &[u8] = b"5629";
+        map.save(&mut store, pk, &data3).unwrap();
+
+        let data4 = Data {
+            name: "Maria Luisa".to_string(),
+            age: 43,
+        };
+        let pk: &[u8] = b"5630";
+        map.save(&mut store, pk, &data4).unwrap();
+
+        // a duplication
+        let pk: &[u8] = b"5631";
+        map.save(&mut store, pk, &data3).unwrap();
+
+        let marias: StdResult<Vec<_>> = map
+            .idx
+            .name_age
```

Comment:
Okay, this is where you are having trouble with the API

Reply:
Yes.